File: backend/src/Workora.Application/Features/Recruitment/DTOs/RecruitmentDto.cs
using Workora.Domain.Enums;

namespace Workora.Application.Features.Recruitment.DTOs;

/// <summary>
/// DTO representing a job vacancy posting.
/// </summary>
public record JobPostingDto(
    int Id,
    Guid Uuid,
    int CompanyId,
    int DepartmentId,
    string? DepartmentName,
    string Title,
    string Description,
    string Requirements,
    EmploymentType EmploymentType,
    string Location,
    int ExperienceYearsMin,
    int ExperienceYearsMax,
    decimal? SalaryMin,
    decimal? SalaryMax,
    JobStatus Status,
    DateOnly? ClosingDate,
    int ApplicantsCount,
    bool IsActive,
    DateTimeOffset CreatedAt);

/// <summary>
/// DTO representing an applicant candidate.
/// </summary>
public record CandidateDto(
    int Id,
    Guid Uuid,
    int JobPostingId,
    string? JobTitle,
    string FirstName,
    string LastName,
    string FullName,
    string Email,
    string? Phone,
    string? ResumeUrl,
    CandidateStage Stage,
    string? RejectionReason,
    DateTimeOffset AppliedDate);

/// <summary>
/// DTO representing a detailed candidate profile with interview rounds and offers.
/// </summary>
public record CandidateDetailDto(
    int Id,
    Guid Uuid,
    int JobPostingId,
    string? JobTitle,
    string FirstName,
    string LastName,
    string FullName,
    string Email,
    string? Phone,
    string? ResumeUrl,
    CandidateStage Stage,
    string? RejectionReason,
    DateTimeOffset AppliedDate,
    IReadOnlyList<InterviewDto> Interviews,
    IReadOnlyList<JobOfferDto> Offers);

/// <summary>
/// DTO representing an interview round.
/// </summary>
public record InterviewDto(
    int Id,
    Guid Uuid,
    int CandidateId,
    string? CandidateName,
    int InterviewerEmployeeId,
    string? InterviewerName,
    DateTimeOffset ScheduledAt,
    string LocationOrLink,
    InterviewStatus Status,
    string? Feedback,
    int? Rating,
    DateTimeOffset? ConductedAt);

/// <summary>
/// DTO representing an employment job offer.
/// </summary>
public record JobOfferDto(
    int Id,
    Guid Uuid,
    int CandidateId,
    string? CandidateName,
    decimal OfferedSalary,
    DateOnly JoiningDate,
    DateOnly ExpiryDate,
    OfferStatus Status,
    DateTimeOffset? SentAt,
    DateTimeOffset? RespondedAt,
    string? Notes);

/// <summary>
/// Recruitment pipeline stage summary DTO.
/// </summary>
public record PipelineStageMetricsDto(
    CandidateStage Stage,
    int Count);

/// <summary>
/// Recruitment pipeline overview response DTO.
/// </summary>
public record RecruitmentPipelineDto(
    IReadOnlyList<PipelineStageMetricsDto> Stages,
    int TotalCandidates);

/// <summary>
/// Request payload for creating a job opening.
/// </summary>
public record CreateJobPostingRequestDto(
    int CompanyId,
    int DepartmentId,
    string Title,
    string Description,
    string Requirements,
    EmploymentType EmploymentType,
    string Location,
    int ExperienceYearsMin,
    int ExperienceYearsMax,
    decimal? SalaryMin,
    decimal? SalaryMax,
    DateOnly? ClosingDate);

/// <summary>
/// Request payload for updating a job opening.
/// </summary>
public record UpdateJobPostingRequestDto(
    int DepartmentId,
    string Title,
    string Description,
    string Requirements,
    EmploymentType EmploymentType,
    string Location,
    int ExperienceYearsMin,
    int ExperienceYearsMax,
    decimal? SalaryMin,
    decimal? SalaryMax,
    DateOnly? ClosingDate);

/// <summary>
/// Request payload for creating an applicant candidate.
/// </summary>
public record CreateCandidateRequestDto(
    int JobPostingId,
    string FirstName,
    string LastName,
    string Email,
    string? Phone,
    string? ResumeUrl);

/// <summary>
/// Request payload for advancing a candidate's stage.
/// </summary>
public record MoveCandidateStageRequestDto(
    CandidateStage Stage);

/// <summary>
/// Request payload for rejecting a candidate.
/// </summary>
public record RejectCandidateRequestDto(
    string? Reason);

/// <summary>
/// Request payload for scheduling an interview.
/// </summary>
public record ScheduleInterviewRequestDto(
    int CandidateId,
    int InterviewerEmployeeId,
    DateTimeOffset ScheduledAt,
    string LocationOrLink);

/// <summary>
/// Request payload for submitting interview feedback.
/// </summary>
public record SubmitInterviewFeedbackRequestDto(
    string Feedback,
    int Rating);

/// <summary>
/// Request payload for generating a job offer.
/// </summary>
public record CreateJobOfferRequestDto(
    int CandidateId,
    decimal OfferedSalary,
    DateOnly JoiningDate,
    DateOnly ExpiryDate,
    string? Notes);

File: frontend/workora-ui/src/app/data/dtos/recruitment.dto.ts
export interface JobPostingDto {
  id: number;
  uuid: string;
  companyId: number;
  departmentId: number;
  departmentName?: string | null;
  title: string;
  description: string;
  requirements: string;
  employmentType: string;
  location: string;
  experienceYearsMin: number;
  experienceYearsMax: number;
  salaryMin?: number | null;
  salaryMax?: number | null;
  status: string;
  closingDate?: string | null;
  applicantsCount: number;
  isActive: boolean;
  createdAt: string;
}

export interface CandidateDto {
  id: number;
  uuid: string;
  jobPostingId: number;
  jobTitle?: string | null;
  firstName: string;
  lastName: string;
  fullName: string;
  email: string;
  phone?: string | null;
  resumeUrl?: string | null;
  stage: string;
  rejectionReason?: string | null;
  appliedDate: string;
}

export interface InterviewDto {
  id: number;
  uuid: string;
  candidateId: number;
  candidateName?: string | null;
  interviewerEmployeeId: number;
  interviewerName?: string | null;
  scheduledAt: string;
  locationOrLink: string;
  status: string;
  feedback?: string | null;
  rating?: number | null;
  conductedAt?: string | null;
}

export interface JobOfferDto {
  id: number;
  uuid: string;
  candidateId: number;
  candidateName?: string | null;
  offeredSalary: number;
  joiningDate: string;
  expiryDate: string;
  status: string;
  sentAt?: string | null;
  respondedAt?: string | null;
  notes?: string | null;
}

export interface CandidateDetailDto extends CandidateDto {
  interviews: InterviewDto[];
  offers: JobOfferDto[];
}

export interface PipelineStageMetricsDto {
  stage: string;
  count: number;
}

export interface RecruitmentPipelineDto {
  stages: PipelineStageMetricsDto[];
  totalCandidates: number;
}

export interface CreateJobPostingRequestDto {
  companyId: number;
  departmentId: number;
  title: string;
  description: string;
  requirements: string;
  employmentType: string;
  location: string;
  experienceYearsMin: number;
  experienceYearsMax: number;
  salaryMin?: number | null;
  salaryMax?: number | null;
  closingDate?: string | null;
}

export interface UpdateJobPostingRequestDto {
  departmentId: number;
  title: string;
  description: string;
  requirements: string;
  employmentType: string;
  location: string;
  experienceYearsMin: number;
  experienceYearsMax: number;
  salaryMin?: number | null;
  salaryMax?: number | null;
  closingDate?: string | null;
}

export interface CreateCandidateRequestDto {
  jobPostingId: number;
  firstName: string;
  lastName: string;
  email: string;
  phone?: string | null;
  resumeUrl?: string | null;
}

export interface MoveCandidateStageRequestDto {
  stage: string;
}

export interface RejectCandidateRequestDto {
  reason?: string | null;
}

export interface ScheduleInterviewRequestDto {
  candidateId: number;
  interviewerEmployeeId: number;
  scheduledAt: string;
  locationOrLink: string;
}

export interface SubmitInterviewFeedbackRequestDto {
  feedback: string;
  rating: number;
}

export interface CreateJobOfferRequestDto {
  candidateId: number;
  offeredSalary: number;
  joiningDate: string;
  expiryDate: string;
  notes?: string | null;
}
